src/main/java/net/sf/jabref/logic/importer/util/MetaDataParser.java

```diff
@@ -0,0 +1,90 @@
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import net.sf.jabref.logic.exporter.FieldFormatterCleanups;
+import net.sf.jabref.logic.l10n.Localization;
+import net.sf.jabref.model.MetaData;
+import net.sf.jabref.model.ParseException;
+import net.sf.jabref.model.groups.GroupTreeNode;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class MetaDataParser {
+
+    private static final Log LOGGER = LogFactory.getLog(MetaDataParser.class);
+
+    public static MetaData parse(Map<String, String> data, String keywordSeparator) throws ParseException {
+        MetaData metaData = new MetaData();
+        metaData.setParsedData(getParsedData(data, keywordSeparator, metaData, Localization.lang("All entries")));
+        return metaData;
+    }
+
+    public static Map<String, List<String>> getParsedData(Map<String, String> inData, String keywordSeparator,
+            MetaData metaData, String allEntriesName)
+            throws ParseException {
+        final Map<String, List<String>> newMetaData = new HashMap<>();
+        for (Map.Entry<String, String> entry : inData.entrySet()) {
+            StringReader data = new StringReader(entry.getValue());
+            List<String> orderedData = new ArrayList<>();
+            // We must allow for ; and \ in escape sequences.
+            try {
+                Optional<String> unit;
+                while ((unit = getNextUnit(data)).isPresent()) {
+                    orderedData.add(unit.get());
+                }
+            } catch (IOException ex) {
+                LOGGER.error("Weird error while parsing meta data.", ex);
+            }
+            if (MetaData.GROUPSTREE.equals(entry.getKey())) {
+                try {
+                    metaData.setGroups(GroupTreeNode.parse(orderedData, keywordSeparator, allEntriesName));
+                // the keys "groupsversion" and "groups" were used in JabRef versions around 1.3, we will not support them anymore
+                } catch (ParseException e) {
+                    throw new ParseException(
+                            Localization
+                                    .lang("Group tree could not be parsed. If you save the BibTeX database, all groups will be lost."),
+                            e);
+                }
+            } else if (MetaData.SAVE_ACTIONS.equals(entry.getKey())) {
+                newMetaData.put(MetaData.SAVE_ACTIONS, FieldFormatterCleanups.parse(orderedData).getAsStringList()); // Without MetaDataChangedEvent
+            } else {
+                newMetaData.put(entry.getKey(), orderedData);
+            }
+        }
+        return newMetaData;
+    }
+
+    /**
+     * Reads the next unit. Units are delimited by ';'.
+     */
+    private static Optional<String> getNextUnit(Reader reader) throws IOException {
+        int c;
```

Comment:
Why not declare it as char? Char ist auto represented as int, so no need to convert it back and forth